Sandrine's Cluedo mystery picker

MindstormingAB opened the pull request:
For this Cluedo project, I first followed the different steps described in the brief to reach the blue level. Thanks to the hints left in the original code, it felt pretty clear so I decided to challenge myself with red and black levels.

I struggled a lot with scope and bad initialization timing but, at then end, I managed to solve everything and I'm satisfied with the final result:
- responsive design
- a shuffleFavouriteWeapon function
- loading animations
- changing the odd of the weapon and the room randomizers
- displaying the information in the DOM
- saving all randomized information in the mystery object.

Link to the deployed project: https://sandrinescluedo.netlify.app/

MindstormingAB committed:
step 1 done

MindstormingAB committed:
Step 1 done inclusive room

MindstormingAB committed:
Step 2 done

MindstormingAB committed:
Step 3 done

MindstormingAB committed:
Step 4 done

MindstormingAB committed:
Adding variables, validation and reset button

MindstormingAB committed:
Created favouriteWeapon and shuffleFavouriteWeapon

MindstormingAB committed:
Animations

MindstormingAB committed:
With higher probability to pick favourite weapon

MindstormingAB committed:
Increased room probability

MindstormingAB committed:
Styling step 1

MindstormingAB committed:
Styling for mobile device

MindstormingAB committed:
Styling for tablet and desktop

MindstormingAB committed:
Fixing main width

MindstormingAB committed:
Adding weapon's favourite room

MindstormingAB committed:
Samll fixes

MindstormingAB committed:
Updated ReadMe

sofiavazs commented on README.md:
```diff
@@ -1,13 +1,22 @@
+
+After Wednesday's lecture with Van, it felt pretty clear how to build the shuffleFavouriteWeapon function but it required a little bit more time to make the loading animations works as I attended.
+
+And then I decided to go for the black level... I spent my Saturday on it, facing several issues related to scope and bad initialization timing, as illustrated in my StackOverflow question (https://stackoverflow.com/c/technigo/questions/1545). After hours of restructuring and control.log, I managed to manipulate the odd in the logic behind generating weapon. Then it was much easier to replicate t ...
```
You did a great job! Awesome!👏🏻

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+const killerLoader = document.getElementById('killerLoader');
+const weaponLoader = document.getElementById('weaponLoader');
+const roomLoader = document.getElementById('roomLoader');
+
```
Nice way of structuring by organising all the variables in one place! Need to try that myself! 😊

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+    weaponImage.style.display = "none";
+    weaponFavouriteRoom.innerHTML = '';
+    revealedMystery.innerHTML = '';
+    setTimeout(function () {
```
oh no! A `function` 🚫! Arrow functions are the way => {😬}

MindstormingAB replied:
@sofiavazs I found a description of the setTimeOut method on w3schools (https://www.w3schools.com/jsref/met_win_settimeout.asp) and I did not dare to change it 😬. But now, I did it, thanks to your comment! 💐0

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+    roomName.innerHTML = '';
+    roomImage.src = '';
+    roomImage.style.display = "none";
+    setTimeout(function () {
```
same as above regarding `function` 🙈

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+
+const pickFavouriteRoom = (weapon) => {
+  const newArrayOfRooms = rooms.map(a => a.name);
+  console.log(newArrayOfRooms);
+  const randomRooms = [
+    ...newArrayOfRooms,
+    ...favouriteRoomDuplicated(weapon.favouriteRoom, 30)
+  ];
+  console.log(favouriteRoomDuplicated(weapon.favouriteRoom, 30));
+  console.log(randomRooms);
+  mystery.pickedRoom = randomSelector(randomRooms);
+  const mysteryRoom = rooms.find((item) => {
+    if (item.name === mystery.pickedRoom) {
+      return item;
+    };
+  });
+  console.log(mysteryRoom);
```
console.logs are awesome in development but not to be left behind 😇

MindstormingAB replied:
Argh! I deleted the ones I commented out and left the visible ones! Thanks for pointing that out, I'm gonna take care of them 🚮

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+
+// Functions related to revelation of mystery 
+
+const revealMystery = () => {
+  if (mystery.killer === undefined || mystery.weapon === undefined || mystery.room === undefined) {
+    revealedMystery.innerHTML = ('Make sure you have picked a killer, a weapon and a room!');
+  } else {
+    revealedMystery.innerHTML = (`The murder was committed by ${mystery.killer.firstName} ${mystery.killer.lastName} with the ${mystery.weapon.name} in the ${mystery.room.name}.`);
+  }
+};
```
Really nice that you made sure the user gets prompted to pick cards and to pick the killer first! 😎

HIPPIEKICK replied:
Could also be written the other way around:
```
if (mystery.killer && mystery.weapon && mystery.room) {
  //Reveal the mystery
} else {
  //Pick some cards first
}
```

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+const resetMystery = () => {
+  location.reload();
+};
 
```
Nice touch with the "reset" button as well!

MindstormingAB replied:
I hope this lazy solution will be accepted by the code teacher 😎

HIPPIEKICK replied:
Lazy is easy ;)

sofiavazs commented on code/index.html:
```diff
@@ -1,37 +1,87 @@
-        <p>KILLER</p>
+        <p class="card-headers">KILLER</p>
+        <div class="loader" id="killerLoader">
+          <i class="fa fa-circle-o-notch fa-spin"></i>
```
Great that you used an icon tag for the loader!

MindstormingAB replied:
Kind of a quick fix but I can recommend it, it is very easy to use. I first used padding to place it in the middle of the card and it spun in a very comical way 😂, margin works better.

sofiavazs commented on code/index.html:
```diff
@@ -1,37 +1,87 @@
+  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+  <meta http-equiv="X-UA-Compatible" content="ie=edge" />
+  <link rel="stylesheet" href="style.css" />
+  <title>🔍 Sandrine's Cluedo</title>
```
Cool touch! 🕵️‍♀️

sofiavazs commented on code/clue.js:
```diff
@@ -1,59 +1,464 @@
+
+// Actions related to Room deck
+
+const favouriteRoomAnimation = () => {
+  if (mystery.killer === undefined) {
+    revealedMystery.innerHTML = ('Pick a killer first!');
+  } else if (mystery.weapon === undefined) {
+    revealedMystery.innerHTML = ('Pick a weapon first!');
+  } else {
```
Awesome job with the user text message prompts! 👏🏻

HIPPIEKICK replied:
This could be written like:
`if (!mystery.killer)`

sofiavazs reviewed (changes requested):
I just want to say that you did such a great job Sandrine! Your project looks great and works as intended.
The code is easily readable, organised and nicely structured it was great to review it and I learned a lot too!
I loved the way you added these little touches throughout the project, like little messages as prompts to the user, the reset and the shuffle favourite weapon button, as well as, making the whole page responsive! Really impressive! 👏🏻
I can tell that you've put a lot of effort to achieve the stretch goals and that doesn't go unnoticed, keep it up!  💪

Super tiny things to fix: 
- `function` to an arrow function. example: `const this = () => {}`
-  Remove console.logs from code 

These are really tiny things but I think it's important to highlight them (since they're in our code guidelines), otherwise everything looked really impressive! 👏🏻

MindstormingAB committed:
Adjustments after Sofia's code review

HIPPIEKICK commented on code/clue.js:
```diff
@@ -1,59 +1,460 @@
+  pistol
+]
+
+const rooms = [
```
Super nice with some room objects as well! 💯

HIPPIEKICK commented on code/clue.js:
```diff
@@ -1,59 +1,460 @@
+  };
+};
+
+const favouriteWeaponDuplicated = (elementToPopulate, howManyTimes) => {
+  let result = [];
+  for (let i = 0; i < howManyTimes; i++) {
+    result.push(elementToPopulate);
+  };
+  return result;
+};
```
🤩

HIPPIEKICK commented on code/clue.js:
```diff
@@ -1,59 +1,460 @@
+  });
+  mystery.weapon = mysteryWeapon;
+  mystery.favouriteRoom = mysteryWeapon.favouriteRoom;
+  weaponLoader.style.display = "none";
+  weaponCard.style.background = "#808080";
+  weaponName.innerHTML = mystery.weapon.name;
+  weaponWeight.innerHTML = `Weight: ${mystery.weapon.weight} oz`;
+  weaponImage.style.display = "inline-block";
```
Consider adding a class instead of styling in the JS-file

MindstormingAB replied:
I created a new class `.weapon-deck-clicked` and replaced `weaponCard.style.background = "#808080";` by `weaponCard.classList.add('weapon-deck-clicked');`
Was it what you meant?

HIPPIEKICK replied:
Yeah, so you can have all the styling in the CSS and then just adding classes in JS 👍

HIPPIEKICK commented on code/clue.js:
```diff
@@ -1,59 +1,460 @@
+// Mystery storage
+
+const mystery = {
+  killer: undefined,
+  favouriteWeapon: undefined,
+  pickedWeapon: undefined,
+  weapon: undefined,
+  favouriteRoom: undefined,
+  pickedRoom: undefined,
+  room: undefined,
```
It's even more semantic approach to initialise mystery object with null values like this :

```
const mystery = {
  killer: null,
  favouriteWeapon: null,
  pickedWeapon: null,
  weapon: null,
  favouriteRoom: null,
  pickedRoom: null,
  room: null,
};
```

HIPPIEKICK reviewed (changes requested):
@MindstormingAB - Be consistent with the semicolons, either you have them everywhere or you don't have them at all.

@sofiavazs Keep it up with the constructive comments 👍

MindstormingAB committed:
Changes after Matilda's feedback

MindstormingAB commented:
@HIPPIEKICK  Thank you for your feedback.
I made the following changes:
- I added semicolons everywhere 😁
- I improved the if statements
- I added classes instead of styling in the JS-file
- I replaced `undefined` with `null`

HIPPIEKICK commented:
> @HIPPIEKICK Thank you for your feedback.
> I made the following changes:
> 
> * I added semicolons everywhere 😁
> * I improved the if statements
> * I added classes instead of styling in the JS-file
> * I replaced `undefined` with `null`

Some semicolons missing after the arrays, no? ;)
